pkg/plugins/golang/options.go

```diff
@@ -0,0 +1,241 @@
+	Namespaced bool
+
+	// Flags that define which parts should be scaffolded
+	DoAPI        bool
+	DoController bool
+	DoDefaulting bool
+	DoValidation bool
+	DoConversion bool
+}
+
+// Validate verifies that all the fields have valid values
+func (opts Options) Validate() error {
+	// Check that the required flags did not get a flag as their value
+	// We can safely look for a '-' as the first char as none of the fields accepts it
+	// NOTE: We must do this for all the required flags first or we may output the wrong
+	// error as flags may seem to be missing because Cobra assigned them to another flag.
+	if strings.HasPrefix(opts.Group, "-") {
+		return fmt.Errorf(groupPresent)
+	}
+	if strings.HasPrefix(opts.Version, "-") {
+		return fmt.Errorf(versionPresent)
+	}
+	if strings.HasPrefix(opts.Kind, "-") {
+		return fmt.Errorf(kindPresent)
+	}
+	// Now we can check that all the required flags are not empty
+	if len(opts.Version) == 0 {
+		return fmt.Errorf(versionRequired)
+	}
+	if len(opts.Kind) == 0 {
+		return fmt.Errorf(kindRequired)
+	}
+
+	// Check if the qualified group has a valid DNS1123 subdomain value
+	if err := validation.IsDNS1123Subdomain(opts.QualifiedGroup()); err != nil {
+		return fmt.Errorf("either group or domain is invalid: (%v)", err)
+	}
+
+	// Check if the version follows the valid pattern
+	if !versionRegex.MatchString(opts.Version) {
+		return fmt.Errorf("version must match %s (was %s)", versionPattern, opts.Version)
+	}
+
+	validationErrors := []string{}
+
+	// Require kind to start with an uppercase character
+	if string(opts.Kind[0]) == strings.ToLower(string(opts.Kind[0])) {
+		validationErrors = append(validationErrors, "kind must start with an uppercase character")
+	}
+
+	validationErrors = append(validationErrors, validation.IsDNS1035Label(strings.ToLower(opts.Kind))...)
+
+	if len(validationErrors) != 0 {
+		return fmt.Errorf("invalid Kind: %#v", validationErrors)
+	}
+
+	// TODO: validate plural strings if provided
+
+	// Ensure apiVersions for k8s types are empty or valid.
+	for typ, apiVersion := range map[string]string{
+		"CRD":     opts.CRDVersion,
+		"Webhook": opts.WebhookVersion,
+	} {
+		switch apiVersion {
+		case "", "v1", "v1beta1":
+		default:
+			return fmt.Errorf("%s version must be one of: v1, v1beta1", typ)
+		}
+	}
+
+	return nil
+}
+
+// QualifiedGroup returns the fully qualified group name with the available information.
+func (opts Options) QualifiedGroup() string {
+	switch "" {
+	case opts.Domain:
+		return opts.Group
+	case opts.Group:
+		return opts.Domain
+	default:
+		return fmt.Sprintf("%s.%s", opts.Group, opts.Domain)
+	}
+}
+
+// GVK returns the GVK identifier of a resource.
+func (opts Options) GVK() resource.GVK {
+	return resource.GVK{
+		Group:   opts.Group,
+		Domain:  opts.Domain,
+		Version: opts.Version,
+		Kind:    opts.Kind,
+	}
+}
+
+// NewResource creates a new resource from the options
+func (opts Options) NewResource(c newconfig.Config) resource.Resource {
```

Comment:
The only way to create a resource now and validate it is by using the Options. 
However, it is specific for Golang and we need these actions for any kind of plugin. (e.g ansible/helm will create resources and then required to validate them according to the kb rules)

In this way, could you add a new resource for the Resource struct and centralize its validation there. The options can use it, however, who is better to know if is valid or not than itself? In this way, IMO the Validate of a resource needs still in/with the resource.

Reply:
That is intentional. `Resource` is exported, so any plugin/third-party library can create them. How we transform our flags into our resource is part of the go plugin implementation. Additionally, if you are doing a light wrap around the go plugin as SDK does, you can also use `Options` as it is exported to.

Reply:
Did you change that? How can I create a newResource and Validate its Resource?

Reply:
The different `Options` struct are fit to a certain plugin flags, so it is not part of the resource model.

A global valid-for-all `Options` field is not a good design decission as it will be really hard to mantain. We would have to add additional fields as required by external plugins, and this changes will block the external plugins. We would have to keep conditionals for different versions and plugins all over the struct methods, which was one of the main reasons we moved to a `Config` interface with independent implementations for each version.